site/src/contexts/ProxyContext.tsx

```diff
@@ -68,39 +114,55 @@ export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
     queryFn: getWorkspaceProxies,
     // This onSuccess ensures the local storage is synchronized with the
     // proxies returned by coderd. If the selected proxy is not in the list,
-    // then the user selection is removed.
+    // then the user selection is ignored.
     onSuccess: (resp) => {
-      setAndSaveProxy(proxy.selectedProxy, resp.regions)
+      // Always pass in the user's choice.
+      setAndSaveProxy(savedProxy, resp.regions)
     },
   })
 
-  // Everytime we get a new proxiesResponse, update the latency check
+  // Every time we get a new proxiesResponse, update the latency check
   // to each workspace proxy.
   const proxyLatencies = useProxyLatency(proxiesResp)
 
-  const setAndSaveProxy = (
-    selectedProxy?: Region,
-    // By default the proxies come from the api call above.
-    // Allow the caller to override this if they have a more up
-    // to date list of proxies.
-    proxies: Region[] = proxiesResp?.regions || [],
-  ) => {
-    if (!proxies) {
-      throw new Error(
-        "proxies are not yet loaded, so selecting a proxy makes no sense. How did you get here?",
+  // If the proxies change or latencies change, we need to update the
+  // callback function.
+  const setAndSaveProxy = useCallback(
+    (
+      selectedProxy?: Region,
+      // By default the proxies come from the api call above.
+      // Allow the caller to override this if they have a more up
+      // to date list of proxies.
+      proxies: Region[] = proxiesResp?.regions || [],
+    ) => {
+      if (!proxies) {
+        throw new Error(
+          "proxies are not yet loaded, so selecting a proxy makes no sense. How did you get here?",
+        )
+      }
+
+      // The preferred proxy attempts to use the user's selection if it is valid.
+      const preferred = getPreferredProxy(
+        proxies,
+        selectedProxy,
+        proxyLatencies,
       )
-    }
-    const preferred = getPreferredProxy(proxies, selectedProxy)
-    // Save to local storage to persist the user's preference across reloads
-    // and other tabs.
-    savePreferredProxy(preferred)
-    // Set the state for the current context.
-    setProxy(preferred)
-  }
+      // Set the state for the current context.
+      setProxy(preferred)
+    },
+    [proxiesResp, proxyLatencies],
+  )
+
+  // This useEffect ensures the proxy to be used is updated whenever the state changes.
+  // This includes proxies being loaded, latencies being calculated, and the user selecting a proxy.
+  useEffect(() => {
+    setAndSaveProxy(savedProxy)
+  }, [savedProxy, proxiesResp, proxyLatencies, setAndSaveProxy])
```

Comment:
Not sure if I'm getting this correctly but let's say the latency changes while I'm on the page, would the proxy be changed automatically?

Reply:
Yes it would.